%gall refactor

jtobin opened the pull request:
This is a general, mostly stylistic refactoring of `%gall`.  I've made an effort to stress clarity and readability throughout; comparatively little structural refactoring has been done, although I've eliminated or created new arms for the `+mo` and `+ap` cores here and there.  I've added some minor commentary where appropriate.

(There are also some related small changes to `%zuse`, `%dill`, and `hood`, namely around the removal of some superfluous types.)

The size of the vane is greatly increased, but this is primarily due to "stretching out" the very terse code that has historically been present.  The code has become dumber, not more complex, and hopefully without having become too verbose in the process (some preliminary review of the style has been encouraging).

There is a lot more that could be done here, but it would be good to get this chunk of work merged so that future contributions can be more appropriately surgical.

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
-    ::::::::::::::::::::::::::::::::::::::::::::::::::::::  vane header
-    ::::::::::::::::::::::::::::::::::::::::::::::::::::::
+::
+::  (rest of arvo)
```
What does this comment mean?  If it's referring to the compile-time subject, it's really Zuse, not just Arvo.

I think we should just delete it.  Every Hoon file has a compilation subject, and all vanes have Zuse, so it's not very informative.

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+::
+|%
+::
+::  +internal-gift: synonym for +cuft.
```
These should be `+$`, since they're types.  This is also a good candidate for shortening.  The style-guide style would suggest something like this:
```
::  $internal-gift: synonym for $cuft
::  $internal-task: synonym for $cush
::  $agent-action: synonym for $club
::
+$  internal-gift  cuft
+$  internal-task  cush
+$  agent-action   club
```

Also, the comments bother me; I think it's because they just repeat the code itself.  I think the comments should be removed or replaced by something explanatory.

I kinda feel like these should be `=*  internal-gift  cuft`, which would emphasize the fact that they're just aliases.

loganallenc replied:
+1 for adding something that is actually explanatory

belisarius222 commented:
I'm leaving some stylistic comments in a stream-of-consciousness manner.  I'll make an actual review comment if I think it should block merging; otherwise, consider these comments advisory.

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+      %cay
+  ==
+::
+::  +volt: voltage.
```
another repetitive comment ... if I see more of these, I will contain my ire and refrain from mentioning them; you know how I feel

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+::
+++  reverse-ames
+  $%
+      :: diff
```
one more space

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+::  +reverse-ames: reverse ames message.
+::
+++  reverse-ames
+  $%
```
I don't like the dangling $%.  It's standard practice to start the contents on the same line as the rune.

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+  $%
+      :: message
+      ::
+      [action=%m mark=mark noun=*]
```
for consistency, we should use the `=mark` syntax here, and in other type definitions

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+  ==
+--
+::
+::  (local arvo)
```
bewildering... what could it mean?  I'm guessing also not your fault

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+::  +internal-note: +ap note.
+::
+++  internal-note
+  $%  [task=%meta =term =vase]
```
It's interesting you put a face on the tag.  I haven't seen this before.  The usual idiom is to `?-  -.note`, which I consider a good use of lark.  The fact that the tag is in the head is actually important to the compiler, so that lark can be thought of as a standard name.  Otherwise everybody has to define their own `task` face or whatever, which is less immediately grokkable.

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+::  duct.
+::
+++  internal-move
+  $:
```
should be
```
$:  =bone
    move=(wind internal-note internal-gift)
==
```
or
```
[=bone move=(wind internal-note internal-gift)]
```
I have a mild preference for the latter.

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+::
+::  +move: Arvo-level move.
+::
+++  move  (pair duct (wind note-arvo gift-arvo))
```
It's weird to have one of these move types be a `pair` and the other have a named `bone` in the head.  We should pick one for this file.

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+::
+::  +gall-old: upgrade path.
+::
+++  gall-old  ?(gall)
```
whoa, I guess this is ok, if a little irregular ... I think this pattern won't lead to the generation of efficient molds, because the compiler would need a `$%`, although maybe that doesn't matter here

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+  $:
+      :: incoming subscribers
+      ::
+      incoming=bitt
```
it would be nice to have an alias for `bitt` and `boat` to maintain consistency of the naming conventions

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+::
+::  +gall: all state.
+::
+++  gall
```
I don't like that this shadows the `gall` defined in Zuse.  I'd be ok with `gall-state` or `state`, but the shadowing will trip people up if they try to modify this code.

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+      blocked=(map term blocked)
+  ==
+::
+::  +routes: new cuff.
```
the use of a plural noun for a type squicks me out, but maybe I could be convinced it's ok

belisarius222 replied:
it breaks my subvocalization: I can't say "this function produces a routes" without displeasure

belisarius222 replied:
do you have a list of routeses? ok I'll stop

loganallenc replied:
"this function produces routes" 

there, I fixed it :)

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+::
+++  gall-payload  +
+::
+::  +mo: Arvo-level move handling.
```
no period at the end of single-line comments like this one

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+      moves=(list move)
+    ==
+  ::
+  ++  mo-state  .
```
We (or maybe just I? I think it's something of a convention) usually use 'state' to refer to a core's data, excluding the battery -- probably excluding the context too.  I got pretty confused when I saw +mo-boot emit a move but just produce 'state'.  I think 'state' usually has the connotation of concrete, clammable data, as distinct from cores, which include code.

This distinction is much sharper in Hoon than in Haskell because everything needs to hot-reload in Urbit and no one knows how to upgrade arbitrary closures, so we want the concrete data separate from the code when we reload.

This is the sort of naming convention we might want to document, although I don't think I knew it existed until just now.

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
-    (~(nest ut p.u.bol) %| -:!>(*bowl))
-  ::  +mo-receive-core: receives an app core built by ford-turbo
+    ::
+    =/  inferred  -:!>(*bowl)
```
I found this variable name misleading, and it's defined a bit far away from its use.  Not egregious, but I think this would actually be clearer stuck into the +nest call, with no name.  You could call it `bowl-type`, but `-:!>` is immediately recognizable to a Hoon programmer as the type spear, so I don't think it's worth naming.

Conversely, I'd try to name `maybe-vase` and `p.u.maybe-vase` a bit more descriptively.  Something like `app-sample-head` or something to convey why we're doing this.

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+    ::
+    =/  =suss  [term %boot now]
+    (mo-give %onto [%.y suss])
+  ::
```
We don't usually include a vertical space above the first line of the comment, although it's not strictly disallowed.  The multiline body should be indented two more spaces, though.

loganallenc commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+::
+++  foreign-response
+  $?  %peer
+      %peel
```
What is a `peel`? I've never seen this before.

joemfb replied:
mark-translated %peer

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+    ::
+    =/  =internal-gift  +>.sign-arvo
+    ::
+    ?-    -.internal-gift
```
`?-` blank comment lines should be at the same indentation level as the rune.

```
?-    -.internal-gift
    %coup
  (mo-give %mack p.internal-gift)
::
    %diff
  ::
  =/  sys-path  [%sys %red t.path]
  =/  =note-arvo  ...
  ...
::
    %quit
  ...
==
```

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+::
+::  +volt: voltage.
+::
+++  volt  ?(%low %high)
```
Apparently this is unused, so it should be ripped out.

belisarius222 commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
-    ?^  app-data
-      ::  update the path
-      ::
-      =.  bum.mas  (~(put by bum.mas) dap u.app-data(by
```
I really hope this doesn't compile

Fang- commented on pkg/arvo/sys/vane/gall.hoon:
```diff
@@ -2,1457 +2,2882 @@
+                attributing.routes.agent-privilege      ::  guest
+                agent-name                              ::  agent
+            ==                                          ::
+            :*  wex=~                                   ::  outgoing
```
@jtobin would tracking/providing these be fairly straightforward for you to implement at this point? If not, can we get a `TODO` marker on this?

jtobin commented:
Thanks for the review lads.  I've addressed almost everything mentioned here locally (including minor structural refactoring around synonyms for things in `%zuse`, etc.) -- I'll give it all a test in aqua and push it as soon as I get the chance.

philipcmonk commented:
> I've addressed almost everything mentioned here locally

Have you pushed this here?  I would like to review your most recent version if this is stale.

jtobin commented:
On Mon, Aug 19, 2019 at 06:31:41PM +0000, Philip Monk wrote:
> Have you pushed this here?  I would like to review your most recent
> version if this is stale.

Just updated the branch so you can take a peek.  I'm going to close this
PR in the meantime until I've run it through aqua and cleaned up these
recent commits; will reopen it then.

ohAitch commented:
Would the recently introduced "mark PR as draft" functionality be perhaps a better fit here?

jtobin commented:
Good point.  *looks for draft button*

matildepark commented:
IIRC you can’t change *to* draft, only submit PRs as drafts.

> On Aug 19, 2019, at 3:12 PM, Jared Tobin <notifications@github.com> wrote:
> 
> Good point. looks for draft button
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jtobin commented:
Crap.  Alright, this one stays closed for now.
